candle-core/src/metal_backend.rs

```diff
@@ -204,33 +215,22 @@ impl MetalDevice {
     /// This method will block the computation because of the
     /// lack of lifetime management through the GPU.
     /// Internal comment for technical details.
-    pub fn new_buffer_with_data<T>(&self, data: &[T]) -> Result<Arc<Buffer>> {
-        let size = core::mem::size_of_val(data) as NSUInteger;
+    pub fn new_buffer_with_data<T>(&self, data: Vec<T>) -> Result<Arc<Buffer>> {
```

Comment:
I'm not sure to understand why passing a `Vec<T>` here is better than the reference. We don't seem to be passing the ownership of the `Vec` to metal as we only extract a pointer out of it, so at the end of the function `data` will be de-allocated which seems unsafe if we don't wait until completion. I'm probably missing something here?

Reply:
Ah, you're right. How about I keep `data` (within MetalDevice) until we know it has been committed/synced?

Reply:
Or maybe we were fine all along 🤔 

[newBufferWithBytes](https://developer.apple.com/documentation/metal/mtldevice/1433429-newbufferwithbytes?language=objc)
> Allocates a new buffer of a given length and initializes its contents by copying existing data into it.
> ...
> Return value: A new [MTLBuffer](https://developer.apple.com/documentation/metal/mtlbuffer?language=objc) instance if the method completes successfully; otherwise nil.

Reply:
For comparison `newBufferWithBytesNoCopy` would be unsafe since it only references the pointer you send in.

Reply:
Cool, can you add a comment in the code with a link to the documentation, that would be great for people looking at it that would wonder why it's safe.